bindings/python/manifold3d.cpp

```diff
@@ -1014,6 +1056,17 @@ NB_MODULE(manifold3d, m) {
           },
           "Returns the vertices of the cross-section's polygons as a "
           "List[List[Tuple[float, float]]].")
+      .def(
+          "to_polygons_numpy",
```

Comment:
This shows my lack of Python knowledge - I think we made `mesh` use numpy, but maybe not the rest? I'd love to understand the tradeoffs better from our Python users; @johnnybigert @axel-angel @zalo, any thoughts? Which APIs could use an update? I am leaning towards a boolean to provide numpy output rather than duplicating everything.

Reply:
I'm also leaning toward `return_numpy` arg over the separate function names. 

As far as which functions next, personally I'd prioritize anywhere that uses open-ended lists/vectors to transfer possibly large numbers of scalars, as these will see the largest performance benefits. Other functions which take small tuples, or x,y,z args are probably not worth messing with - as python gives very clean syntax to convert between ndarray, tuple, and multi-args.

Reply:
And a subtle point is that numpy wants uniform dimensions, so in cases like this one which are a list of variable length polygon points, I have it return a list of ndarrays as a compromise, to make the data significantly _more_ contiguous and compact even if it's not 100% contiguous.

Reply:
I think it’s pretty safe to remove the bool, make NumPy a hard requirement, and to return it for everything all the time.  They’re functionally equivalent to traditional lists, but they add a ton of functionality (and operator overloads for batch math!)  For anything involving any math/numbers, It’s probably the most used/depended on Python package of all time by a wide margin.  It’s one of the few “libraries” that makes everything it touches just faster and better, and every Python user (who isn’t just scraping websites) already has it.

For ragged arrays, a list of numpy arrays is the best compromise.

I can go in and add the other APIs if that makes sense 👍
(I can also come up with a list of extremely popular libraries that also rely entirely on Numpy if you’re not convinced)

Reply:
Some of the other APIs are already numpy-only so it's not really a question of dependency. It's more a question of introducing breaking API changes. The bool arg allows existing users to upgrade without breaking.

I'm a new user of manifold, so I personally have no problem with breaking API changes :P

Reply:
I pushed the `return_numpy` version, since it's better than the previous version at least.

If we do decide to break the APIs to switch fully to numpy, we should try to bundle them all under a clear version bump, so people can pin to old version if not wanting to migrate.

Reply:
> They’re functionally equivalent to traditional lists (so the change won’t break most code!)

Oh I missed this point earlier. That is a good point, many common interactions would still work, for example:

```
x = my_cross_section.to_polygons()
for poly in x:
  for x, y in poly
    # do stuff
```

Reply:
Turns out one of the `run_all.py` test cases breaks with this change, see the new diff in run_all.py for details.
Arguably, creating a `set()` from point tuples is more of a contrived test fixture than a realistic use though.

Reply:
Thanks @zalo, that's a good point about this being a nearly transparent change. And yeah, I'm not worried about that test change, @wrongbad. Honestly, I think I'm comfortable calling this part of v2.3 - I don't think we have quite enough users to be quite that picky about small breaking changes. Does that sound reasonable @pca006132?

Reply:
I think using numpy arrays directly is fine.